cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaListenersReconciler.java

```diff
@@ -0,0 +1,1040 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.cluster.operator.assembly;
+
+import io.fabric8.kubernetes.api.model.Secret;
+import io.fabric8.kubernetes.api.model.Service;
+import io.fabric8.kubernetes.api.model.networking.v1.Ingress;
+import io.fabric8.openshift.api.model.Route;
+import io.strimzi.api.kafka.model.CertAndKeySecretSource;
+import io.strimzi.api.kafka.model.KafkaResources;
+import io.strimzi.api.kafka.model.listener.NodeAddressType;
+import io.strimzi.api.kafka.model.listener.arraylistener.GenericKafkaListener;
+import io.strimzi.api.kafka.model.listener.arraylistener.GenericKafkaListenerConfigurationBroker;
+import io.strimzi.api.kafka.model.status.ListenerAddressBuilder;
+import io.strimzi.api.kafka.model.status.ListenerStatus;
+import io.strimzi.api.kafka.model.status.ListenerStatusBuilder;
+import io.strimzi.operator.PlatformFeaturesAvailability;
+import io.strimzi.operator.cluster.model.Ca;
+import io.strimzi.operator.cluster.model.ClusterCa;
+import io.strimzi.operator.cluster.model.DnsNameGenerator;
+import io.strimzi.operator.cluster.model.InvalidResourceException;
+import io.strimzi.operator.cluster.model.KafkaCluster;
+import io.strimzi.operator.cluster.model.ListenersUtils;
+import io.strimzi.operator.common.Reconciliation;
+import io.strimzi.operator.common.ReconciliationLogger;
+import io.strimzi.operator.common.Util;
+import io.strimzi.operator.common.operator.resource.IngressOperator;
+import io.strimzi.operator.common.operator.resource.IngressV1Beta1Operator;
+import io.strimzi.operator.common.operator.resource.RouteOperator;
+import io.strimzi.operator.common.operator.resource.SecretOperator;
+import io.strimzi.operator.common.operator.resource.ServiceOperator;
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Future;
+
+import java.nio.charset.StandardCharsets;
+import java.security.cert.CertificateEncodingException;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.singletonList;
+
+/**
+ * Class used for reconciliation of Kafka listeners. This class contains both the steps of the Kafka
+ * reconciliation pipeline related to the listeners and is also used to store the state between them.
+ */
+public class KafkaListenersReconciler {
```

Comment:
Trying to figure out why a separate reconciler (from the Kafka one) is needed for the listeners. Just a design choice to make the `KafkaReconciler` less filled of actions or mandatory decision because of any issues?

Reply:
It could have been in one class, yes. But my thoughts were following:
* The combined class would have had over 2000 lines of code, so fairly long
* The listeners reconciliation is fairly long but relatively self-contained, so it can be easily extracted. Yet, its functions are related in terms of functionality and code and rely on each other (e.g. the Routes or Ingresses are created on top of the Services)

So it seemed like a good candidate to split a bit more. 

I also didn't really do a complete refactoring of the unit tests. I really just made them work as they were but testing the new code. But once we get to that as well, there are a many things to test just with the listeners and their different types. So this should also make the test easier in the future because you will test only the smaller class for the listeners and not the main KafkaRecocniler.